---
layout: default
title:  'RNAseq'
---

# <a name="begin"></a> RNA-seq data processing and analysis tutorial
RNA-seq has become a powerful approach to study the continually changing cellular transcriptome. Here, one of the most common questions is to identify genes that are differentially expressed between two conditions, e.g. controls and treatment. The **main** exercise in this tutorial will take you through a basic bioinformatic analysis pipeline to answer just that, it will show you how to find differentially expressed (DE) genes. Briefly,

The **main exercise** is split into four parts: 

   1) Setting up your work area in a good organised, reproducible, way
      * will take about an hour (!)
   2) Checking the quality of your samples - which ones, if any, needs to be excluded? 
      * check the quality of the raw reads with [FastQC](#fastqc) 
   3) Align your samples
      * map the reads to the reference genome using [Star](#star) 
      * convert between SAM and BAM files format using [Samtools](#samtools) 
      * assess the post-alignment reads quality using [QualiMap](#qualimap) 
   4) Determine expressed, and differentially expressed gens
      * count the reads overlapping with genes regions using [featureCounts](#featurecounts) 
      * build statistical model to find DE genes using [edgeR](#descript)

As discussed during the lecture, RNA-seq experiment does not end with a list of DE genes. If you have time after completing the main exercise, try one (or more) of the bonus exercises. The bonus exercises can be run independently of each other, so choose the one that matches your interest.

Additional exercises:  

   * <a href="Bonus_functional_annotations.html">Functional annotation</a> how to put DE genes in biological context 
   * <a href="Bonus_exon_usage.html">Exon usage</a> - how to perform analysis of differential exon usage and study alternative splicing 
   * <a href="Bonus_visualisation.html">Visualisation</a> - how to view RNA-seq bam files and present DE results with graphics 
   * <a href="Bonus_de_novo.html"><i>De novo</i> transcriptome assembly</a> 


# Good practise

Unfortunately a large part of a bioinformaticians life is to ensure you have the right files, from the right genome versions, in the right place in order to run your analysis efficiently. Therefore about the first hour of the lab will be about setting up your work area (workDir) and results directory (resDir) properly. We could have prepared this for you, but then you would not be prepared for some of the most common "challenges".

## Common problems

Most of the problems you will encounter in this lab is different types of typos, therefore look closely at what the error message is saying, and try to solve it.

## Your work area

When doing a project you should always work in the project directory, for the purpose of this lab, you should work in your home directory.

{% highlight bash %}
cd ~/
{% endhighlight %}

## Where to store results

Analysis results should be stored on the nobackup
{% highlight bash %}
mkdir /home/<username>/nobackup/rnaseqlab
export resDir="/home/<username>/nobackup/rnaseqlab"
{% endhighlight %}

The first command creates a place to store all your results, eg a folder called <i>rnaseqlab</i> on the nobackup area.

The second command makes it easier to keep track of your paths, by asking the system to create an environmental variable (see <a href="https://en.wikipedia.org/wiki/Environment_variable">here</a> for a description if you want), this will make it easier for you to work with this lab.

For those of you that will be writing a bash script instead define a parameter at the top of your script file
{% highlight bash %}
resDir="/home/<username>/nobackup/rnaseqlab"
{% endhighlight %}
Then everything else will work exactly the same as the remaining example code


# Data description

The data you will be using in this exercise is from the recent paper [YAP and TAZ control peripheral myelination and the expression of laminin receptors in Schwann cells. Poitelon et al. Nature Neurosci. 2016](http://www.nature.com/neuro/journal/v19/n7/abs/nn.4316.html). In the experiments performed in this study, YAP and TAZ were knocked-down in Schwann cells to study myelination, using the sciatic nerve in mice as a model.

Myelination is essential for nervous system function. Schwann cells interact with neurons and the basal lamina to myelinate axons using receptors, signals and transcription factors. Hippo pathway is an evolutionary conserved pathway involved in cell contact inhibition, and it acts to promote cell proliferation and inhibits apoptosis. The pathway integrates mechanical signals (cell polarity, mechanotransduction, membrane tension) and gene expression response. In addition to its role in organ size control, the Hippo pathway has been implicated in tumorigenesis, for example its deregulation occurs in a broad range of human carcinomas. Transcription co-activators YAP and TAZ are two major downstream effectors of the Hippo pathway, and have redundant roles in transcriptional activation.

The material for RNA-seq was collected from 2 conditions (wt and YAP(kd)TAZ(kd)), each in 3 biological replicates (see table below).

|  Accession  | Condition | Replicate |
| --- | ----------- | --------- |
| SRR3222409 |  KO | 1 |
| SRR3222410 |  KO | 2 |
| SRR3222411 |  KO | 3 |
| SRR3222412 |  WT  | 1 |
| SRR3222413 |  WT  | 2 |
| SRR3222414 |  WT  | 3 |

For the purpose of this tutorial, that is to shorten the time needed to run various bioinformatics steps, we have down-sampled the original files. We randomly sampled, without replacement, 25% reads from each sample, using fastq-sample from the [fastq-tools](http://homes.cs.washington.edu/~dcjones/fastq-tools/) tools.

# Bioinformatics: processing raw sequencing files
Reading manuals, trying different tools/options, finding solutions to problems are daily routine work for bioinformaticians. By now you should have some experience with using command line and various bioinformatic tools, so in order to feel like a pro we encourage you to try your own solutions to the problems below, before checking the solution key. Click to see the suggested answers to compare them with your own solutions. Discuss with person next to you and ask us when in doubt. Remember that there is more than one way to skin a cat. Have fun!

<br />
[Jump to the top](#begin)

## Preparing your working directory
To get going, let's book a node, create a directory named <b>transcriptome</b> in your results directory and link the raw sequencing files .fastq.gz. NB! Remember to replace <username> with your uppmax id, and <courseID> with the code for the course, throughout the exercise.

:computer: **Book a node.** As for other tutorials in this course we have reserved half a node per person. If you have not done it yet today book a node now as otherwise you will take away resources from your fellow course participants.
<details>
<summary>:key: Click to see how to book a node on Thursday</summary>
{% highlight bash %}
salloc -A <courseID> -t 08:00:00 -p core -n 8 --no-shell --reservation=<courseID>_4 &

INTERACTIVE

interactive -A <courseID> -t 04:00:00 -p core -n 8

{% endhighlight %} 
</details>  


:computer: **Create a folder** named <b>transcriptome</b>, then **Create  a sub-folder** called <b>DATA</b>.
<details>
<summary>:key: Click to see suggested commands</summary>
{% highlight bash %}
mkdir $resDir/transcriptome
mkdir $resDir/transcriptome/DATA
{% endhighlight %} 
</details>  
<br />

:computer: **Sym-link** the .fastq.gz files located in <b><i>/sw/courses/ngsintro/rnaseq/DATA/p25</i></b>. :bulb: A great chance to practice your bash loop skills.
<details>
<summary>:key: Click to see suggested commands</summary>
{% highlight bash %}
cd $resDir/transcriptome/DATA
for i in /sw/courses/ngsintro/rnaseq/DATA/p25/*
  do ln -s $i
done
{% endhighlight %} 
</details>  
<br />

<b>What does the above command do?</b> Instead of copying the files from the course directory to your results directory it creates a <a href="https://en.wikipedia.org/wiki/Symbolic_link">symbolic link</a> to the file. You can access the file just as if it would really be in your directory but it will actually not use up any of your file quota, or UppMax storage capacity, by storing the same file in two places.

:white_check_mark: **Check** if you linked the files correctly. You now should be able to see 12 links to the .fastq.gz files.
{% highlight bash %}
ll $resDir/transcriptome/DATA

SRR3222409_1.fastq.gz -> /sw/courses/ngsintro/rnaseq/DATA/p25/SRR3222409_1.fastq.gz
SRR3222409_2.fastq.gz -> /sw/courses/ngsintro/rnaseq/DATA/p25/SRR3222409_2.fastq.gz
...
SRR3222414_2.fastq.gz -> /sw/courses/ngsintro/rnaseq/DATA/p25/SRR3222414_2.fastq.gz
{% endhighlight %}
<br/>
<br />
[Jump to the top](#begin)


## <a name="fastqc"></a> FastQC: quality check of the raw sequencing reads
After receiving raw reads from a high throughput sequencing centre it is essential to check their quality. Why waste your time on data analyses of the poor quality data? FastQC provide a simple way to do some quality control check on raw sequence data. It provides a modular set of analyses which you can use to get a quick impression of whether your data has any problems of which you should be aware before doing any further analysis.

:mag: **Read** more on [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/). Can you figure out how to run it on Uppmax?  
<br />

:computer: **Create** a <i>fastqc</i> folder in your <i>transcriptome</i> directory. **Navigate to fastqc folder**.
<details>
<summary>:key: Click to see suggested commands</summary>
{% highlight bash %}
mkdir $resDir/fastqc
cd $resDir/fastqc
{% endhighlight %} 
</details>  
<br />

:computer: **Load** bioinfo-tools and FastQC modules
<details>
<summary>:key: Click to see suggested commands</summary>
{% highlight bash %} 
module load bioinfo-tools 
module load FastQC/0.11.5
{% endhighlight %} 
</details>  
<br />

:computer: **Run** FastQC on all the .fastq.gz files located in the <i>transcriptome/DATA</i>. **Direct the output** to the  <i>fastqc</i> folder. :bulb: Check the FastQC option for input and output files. :bulb: The bash loop comes handy again.
<details>
<summary>:key: Click to see suggested commands</summary>
{% highlight bash %}
for i in $resDir/transcriptome/DATA/* 
do 
  fastqc $i -o $resDir/transcriptome/fastqc/ 
done
{% endhighlight %}
</details>  
<br />

:mag: **Download** the FastQC for the proceeded sample from Uppmax to your compute and **have a look** at it. **Go back** to the [FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) website and **compare** your report with [Example Report for the Good Illumina Data](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/good_sequence_short_fastqc.html) and [Example Report for the Bad Illumina Data](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/bad_sequence_fastqc.html) data.  
<br />

:open_mouth: Discuss whether you'd be happy when receiving this very data from the sequencing facility.
<br />
<br />
[Jump to the top](#begin)
<br />
<br />


## <a name="star"></a> STAR: aligning reads to a reference genome
After verifying that the quality of the raw sequencing reads is acceptable we can map the reads to the reference genome. There are many mappers/aligners available, so it may be good to choose one that is adequate for your type of data. Here, we will use a software called STAR (Spliced Transcripts Alignment to a Reference) as it is good for generic purposes, fast, easy to use and has been shown to outperform many of the other tools when aligning 2x76bp paired-end data (2012). Before we begin mapping, we need to obtain genome reference sequence (.fasta file) and a corresponding annotation file (.gtf) and build a STAR index. Due to time constrains, we will practice on chromosome 11 only. Then we will use the pre-prepared index for the entire genome to do the actual mapping.

### Accessing reference genome and genome annotation file
It is best if the reference genome (.fasta) and annotation (.gtf) files come from the same source to avoid potential naming conventions problems. It is also good to check in the manual of the aligner you use for hints on what type of files are needed to do the mapping.

:mag: **Check** [STAR](https://github.com/alexdobin/STAR) manual what files are needed for the mapping.  
<br />

:open_mouth: What is the idea behind building STAR index? What files are needed to build one? Where do we take them from? Could one use a STAR index that was generated before?  
<br />

:computer: **Create** the <i>reference</i> sub-folder in the <i>transcriptome</i> directory
<details>
<summary>:key: Click to see how to create the directory </summary>
{% highlight bash %}
mkdir $resDir/transcriptome/reference
{% endhighlight %}
</details>  
<br />

:computer: **Download** the reference genome .fasta file for chromosome 11, mouse and the corresponding genome annotation .gtf file from Ensmeble webite.
<details>
<summary>:key: Click for the link to the Ensembl website </summary>
<a href="http://www.ensembl.org/info/data/ftp/index.html">http://www.ensembl.org/info/data/ftp/index.html</a>
</details>
<details>
<summary>:key: Click to see file names to be downloaded </summary>
{% highlight bash %}
Mus_musculus.GRCm38.dna.chromosome.11.fa
Mus_musculus.GRCm38.85.gtf
{% endhighlight %}
</details>
<br />
<details>
<summary>:key: Click to see how to transfer files from Ensembl website to Uppmax </summary>
{% highlight bash %}
cd $resDir/transcriptome/reference
wget ftp://ftp.ensembl.org/pub/release-85/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.chromosome.11.fa.gz
wget ftp://ftp.ensembl.org/pub/release-85/gtf/mus_musculus/Mus_musculus.GRCm38.85.gtf.gz
{% endhighlight %}
</details>  
<br />

The files you have just downloaded are compressed using gzip; you need to decompress them before use.

<details>
<summary>:key: Click to see how </summary>
{% highlight bash %}
gzip -d *gz
{% endhighlight %}
</details>  
<br />

:white_check_mark: You should now have *Mus\_musculus.GRCm38.dna.chromosome.11.fa* and *Mus\_musculus.GRCm38.85.gtf* in the sub-folder _reference_  
{% highlight bash %}
ll -h $resDir/transcriptome/reference/
-r--r--r-- 1 olga olga 867699172 Sep 14 15:33 Mus_musculus.GRCm38.81.gtf
-rw-rw-r-- 1 olga olga 124117311 Sep 14 15:42 Mus_musculus.GRCm38.dna.chromosome.11.fa
{% endhighlight %}
<br />
<br />
[Jump to the top](#begin)

### preparing index

:computer: **Create <i>indexChr11</i> sub-folder** in the <i>transcriptome</i> directory
<details>
<summary>:key: Click to see how to create directory</summary>
{% highlight bash %}
mkdir $resDir/transcriptome/indexChr11
cd $resDir/transcriptome/indexChr11
{% endhighlight %}
</details>  
<br />

:computer: **Load STAR module** on Uppmax. :bulb: Use <i>module spider star</i> to check which version of STAR are available and load the latest one.
<details>
<summary>:key: Click to see how to load module</summary>
{% highlight bash %}
module spider star
module load star/x.x.x
{% endhighlight %}
</details>  
<br />

:computer: **Build STAR index** for chromosome 11 using the downloaded reference .fasta and gene annotation .gtf files. :bulb: Check <a href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf">STAR manual</a> for details
<details>
<summary>:key: Click again to see suggested commands</summary>
{% highlight bash %}
star --runMode genomeGenerate --runThreadN 8 --genomeDir $resDir/transcriptome/indexChr11 --genomeFastaFiles $resDir/transcriptome/reference/Mus_musculus.GRCm38.dna.chromosome.11.fa --sjdbGTFfile $resDir/transcriptome/reference/Mus_musculus.GRCm38.85.gtf
{% endhighlight %}
</details>
<br />

:white_check_mark: Check if building the index worked, you should at the very least have the below files
{% highlight bash %}
ll -h $resDir/transcriptome/indexChr11
-rw-rw-r-- 1 olga olga   10 Sep 14 15:53 chrLength.txt
-rw-rw-r-- 1 olga olga   13 Sep 14 15:53 chrNameLength.txt
-rw-rw-r-- 1 olga olga    3 Sep 14 15:53 chrName.txt
-rw-rw-r-- 1 olga olga   12 Sep 14 15:53 chrStart.txt
-rw-rw-r-- 1 olga olga  750 Sep 14 15:53 genomeParameters.txt
-rw-rw-r-- 1 olga olga  14K Sep 14 15:54 Log.out
-rw-rw-r-- 1 olga olga  20M Sep 14 15:54 SA_7
drwx------ 2 olga olga 4.0K Sep 14 15:53 _STARtmp
{% endhighlight %}
<br />

:computer: **Sym-link STAR index** to for the entire genome into the <i>transcriptome</i> directory. The index for the whole genome was prepared for us before class in the very same way as for the chromosome 11 in steps above. It just requires more time (ca. 4h) to run. The index can be found here: <b><i>/sw/courses/ngsintro/rnaseq/index</i></b>
<details>
<summary>:key: Click again to see how to link the index</summary>
{% highlight bash %}
cd $resDir/transcriptome/
ln -s /sw/courses/ngsintro/rnaseq/index
{% endhighlight %}
</details>
<br />
<br />
[Jump to the top](#begin)

### Mapping
Now we are ready to map our reads to the reference genome, via STAR index.  
<br />

:computer: **Create _star_ sub-folder** in the <i>transcriptome</i> directory. **Create sub-sub-folder named <i>SRR3222409</i>** to save the mapping results for the sample SRR3222409.
<details>
<summary>:key: Click to see how to create folders </summary>
{% highlight bash %}
mkdir $resDir>/transcriptome/star
mkdir $resDir/transcriptome/star/SRR3222409
{% endhighlight %}
</details>
<br />

:computer: **Map reads** to the reference genome for SRR3222409 sample. Do not forget that we are working with paired-end reads so each sample has two matching reads file. **Check** the <a href="https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf">STAR manual</a> for the parameters to:
<br />
* use index for the entire genome <br />
* to read in zipped .fastq.gz files for both forward and reverse reads <br />
* to run the job on the 8 allocated cores <br />
* to direct the mapping results to the _SRR3222409_ sub-sub folder <br />
* to give the results prefix _SRR3222409_  <br />

<details>
 <summary>:key: Click to see how to write the mapping command with the above parameters</summary>
 {% highlight bash %}
mkdir $resDir/transcriptome/star
mkdir $resDir/transcriptome/star/SRR3222409
star --genomeDir $resDir/transcriptome/index/complete --readFilesIn $resDir/transcriptome/DATA/SRR3222409_1.fastq.gz $resDir/transcriptome/DATA/SRR3222409_2.fastq.gz --runThreadN 8 --readFilesCommand zcat --outFileNamePrefix $resDir/transcriptome/star/SRR3222409/SRR3222409_
 {% endhighlight %}
 </details>
<br />

:white_check_mark: You should now have .sam file in the <i>$resDir/transcriptome/star/SRR3222409</i> folder as well as a series of log files. Have a look how the mapping went.
{% highlight bash %}
ll -h $resDir/transcriptome/star/SRR3222409

total 3.0G
-rw-rw-r-- 1 olga olga 3.0G Sep 14 16:05 SRR3222409_Aligned.out.sam
-rw-rw-r-- 1 olga olga 1.9K Sep 14 16:05 SRR3222409_Log.final.out
-rw-rw-r-- 1 olga olga  18K Sep 14 16:04 SRR3222409_Log.out
-rw-rw-r-- 1 olga olga  364 Sep 14 16:04 SRR3222409_Log.progress.out
-rw-rw-r-- 1 olga olga 3.6M Sep 14 16:05 SRR3222409_SJ.out.tab
drwx------ 2 olga olga 4.0K Sep 14 16:04 SRR3222409__STARtmp
{% endhighlight %}
<br/>
<br/>

:computer: **Map or copy over**. Map the remaining samples in the analogous way.

Hint if you write a bash look, then please note that the underscore (for example before the 1 and 2 to define the forward and reverse strand) will be understood as part of the parameter!
<br />
<br />
[Jump to the top](#begin)
<br />
<br />


### <a name="samtools"></a> Samtools: converting between SAM and BAM
Before we proceed further with our data processing, let's convert our mapped reads from STAR, saved in the default .SAM text format, into the binary .BAM format. Why? BAM files take less space so it is easier to store them and they are the most commonly required file format for many of the down-stream bioinformatics tools. In addition, they can be sorted and indexed shortening the time needed to proceed them in comparison with .SAM format. Also, then they will be ready for exploration in IGV, the Integrative Genomic Viewer.

:mag: **Read** through [Samtools](http://www.htslib.org/doc/samtools.html) documentation and see if you can figure it out how to:
* convert SAM into BAM
* sort BAM files
* index BAM files
<br />

:computer: **Create _bams_ sub-folder** in _transcriptome_, **navigate to _bams_ sub-folder** and **load samtools module**
<details>
<summary>:key: Click to see the suggested commands, file by file</summary>
{% highlight bash %}
mkdir $resDir/transcriptome/bams
cd $resDir/transcriptome/bams 
module spider samtools
module load samtools/x.x
{% endhighlight %}
</details>
<br />


:computer: **Sym-link** in _bams_ sub-folder all the SAM files containing the mapped reads, as created during the Star mapping step. :bulb: You can use the bash loop if you apply wild cards (slightly more advanced but try first before looking at the answer key)
<details>
<summary>:key: Click to see the suggested commands, sample by sample</summary>
{% highlight bash %}
ln -s $resDir/transcriptome/star/SRR3222409/SRR3222409_Aligned.out.sam
{% endhighlight %}
</details>
<details>
<summary>:key: Click to see the suggested commands, using bash loop</summary>
{% highlight bash %}
for i in $resDir/transcriptome/star/**/*.sam 
  do ln -s $i
done
{% endhighlight %}
</details>
<br />

:white_check_mark: Check to see the linked files
{% highlight bash %}
ll -h  $resDir/transcriptome/bams/*.sam

lrwxrwxrwx 1 olga olga 72 Sep 14 16:39 SRR3222409_Aligned.out.sam -> /home/olga/glob/transcriptome/star/SRR3222409/SRR3222409_Aligned.out.sam
lrwxrwxrwx 1 olga olga 72 Sep 14 16:39 SRR3222410_Aligned.out.sam -> /home/olga/glob/transcriptome/star/SRR3222410/SRR3222410_Aligned.out.sam
...
lrwxrwxrwx 1 olga olga 72 Sep 14 16:39 SRR3222414_Aligned.out.sam -> /home/olga/glob/transcriptome/star/SRR3222414/SRR3222414_Aligned.out.sam
{% endhighlight %}
<br/>

:computer: **Convert SAM to BAM**: for the first sample <i>SRR3222409\_Aligned.out.sam</i> into <i>SRR3222409\_Aligned.out.bam</i>
<details>
<summary>:key: Click to see the suggested commands</summary>
{% highlight bash %}
cd $resDir/transcriptome/bams/
samtools view -bS -o SRR3222409_Aligned.out.bam SRR3222409_Aligned.out.sam
{% endhighlight %}
</details>
<br />

:computer: **Convert SAM to BAM** for the remaining samples
<br />

:white_check_mark: Check to see the converted BAM files
{% highlight bash %}
ll -h  $resDir/transcriptome/bams/*.bam

-rw-rw-r-- 1 olga olga 792M Sep 14 16:46 SRR3222409_Aligned.out.bam
-rw-rw-r-- 1 olga olga 596M Sep 14 16:51 SRR3222410_Aligned.out.bam
...
-rw-rw-r-- 1 olga olga 1.1G Sep 14 16:54 SRR3222414_Aligned.out.bam
{% endhighlight %}
<br/>

In general here is where you would <b>remove your SAM files to save space</b>, but DO NOT DO THIS today, the featureCount we use later requires samfiles as input. If you have removed you <i>sam</i> files when you get to that part, then use <i>samtools view *bam</i>to get them back/pipe it directly to most bioinformatics software.

:computer: **Sort BAM file** sort <i>SRR3222409\_Aligned.out.bam</i> file and save it as <i>SRR3222409\_Aligned.out.sorted.bam</i> in the <i>bams</i> sub-folder
<details>
<summary>:key: Click to see how to sort BAM file</summary>
{% highlight bash %}
samtools sort -o $resDir/transcriptome/bams/SRR3222409_Aligned.out.sorted.bam -T sort_temp $resDir/transcriptome/bams/SRR3222409_Aligned.out.bam
{% endhighlight %}
</details>
<br />

:computer: **Sort BAM files** for the remaining samples 

:white_check_mark: Check to see the sorted BAM files
{% highlight bash %}
ll -h  $resDir/<username>/transcriptome/bams/*.sorted.bam

-rw-rw-r-- 1 olga olga 492M Sep 14 16:59 SRR3222409_Aligned.out.sorted.bam
-rw-rw-r-- 1 olga olga 375M Sep 14 17:01 SRR3222410_Aligned.out.sorted.bam
...
-rw-rw-r-- 1 olga olga 631M Sep 14 17:02 SRR3222414_Aligned.out.sorted.bam
{% endhighlight %}
<br/>

:computer: **Index the sorted BAM files**
<details>
<summary>:key: Click to see how to index BAM file, sample by sample</summary>
{% highlight bash %}
samtools index $resDir/transcriptome/bams/SRR3222410_Aligned.out.sorted.bam
{% endhighlight %}
</details>
<details>
<summary>:key: Click to see how to index BAM file, using bash loop</summary>
{% highlight bash %}
for i in  $resDir/transcriptome/bams/*.sorted.bam
  do samtools index $i
done
{% endhighlight %}
</details>
<br />

:white_check_mark: Check to see the BAM indexes
{% highlight bash %}
ll -h  $resDir/transcriptome/bams/*.bai

-rw-rw-r-- 1 olga olga 2.7M Sep 14 17:06 SRR3222409_Aligned.out.sorted.bam.bai
-rw-rw-r-- 1 olga olga 2.7M Sep 14 17:06 SRR3222410_Aligned.out.sorted.bam.bai
...
-rw-rw-r-- 1 olga olga 2.8M Sep 14 17:07 SRR3222414_Aligned.out.sorted.bam.bai
{% endhighlight %}
<br/>
<br />
[Jump to the top](#begin)


### <a name="qualimap"></a> QualiMap: post-alignment quality control
Some important quality aspects, such as saturation of sequencing depth, read distribution between different genomic features or coverage uniformity along transcripts, can be measured only after mapping reads to the reference genome. One of the tools to perform this post-alignment quality control is QualiMap. QualiMap examines sequencing alignment data in SAM/BAM files according to the features of the mapped reads and provides an overall view of the data that helps to the detect biases in the sequencing and/or mapping of the data and eases decision-making for further analysis.

:mag: **Read** through [QuliMap](http://qualimap.bioinfo.cipf.es/doc_html/intro.html) documentation and see if you can figure it out how to run it to assess post-alignment quality on the RNA-seq mapped samples. The tool is already installed on Uppmax and available as QuliMap module  
<br />

:computer: **Create QualiMap** sub-folder in <i>transcriptome</i> directory, **navigate to <i>qualimap</i> sub-folder** and **load QualiMap/2.2 module**
<details>
<summary>:key: Click to see the suggested commands</summary>
{% highlight bash %}
mkdir $resDir/transcriptome/qualimap 
cd $resDir/transcriptome/qualimap/
module spider QualiMap
module load QualiMap/x.x
{% endhighlight %}
</details>
<br />

:computer: **Run QualiMap** for the fist sample on sorted BAM file <i>SRR3222409\_Aligned.out.sorted.bam</i> **directing** the results to <i>$resDir/transcriptome/qualimap/SRR3222409</i> folder. :bulb: QualiMap creates the folder if you specify the right parameter
<details>
<summary>:key: Click to see the suggested commands</summary>
{% highlight bash %}
qualimap rnaseq -pe -bam $resDir/transcriptome/bams/SRR3222409_Aligned.out.sorted.bam -gtf $resDir/transcriptome/reference/Mus_musculus.GRCm38.85.gtf --outdir $resDir/transcriptome/qualimap/SRR3222409 --java-mem-size=63G > /dev/null 2>&1
{% endhighlight %}
</details>
<br />

:white_check_mark: Check if the QualiMap run correctly
{% highlight bash %}
ll -h $resDir/transcriptome/qualimap/SRR3222409/

total 28K
drwxrwxr-x 2 olga olga 4.0K Sep 14 17:24 css
drwxrwxr-x 2 olga olga 4.0K Sep 14 17:24 images_qualimapReport
-rw-rw-r-- 1 olga olga  11K Sep 14 17:24 qualimapReport.html
drwxrwxr-x 2 olga olga 4.0K Sep 14 17:24 raw_data_qualimapReport
-rw-rw-r-- 1 olga olga 1.2K Sep 14 17:24 rnaseq_qc_results.txt
{% endhighlight %}
<br/>


:computer: **Run QualiMap** for the remaining sorted BAM files or **copy the results over**. These can be found <i>/sw/courses/ngsintro/rnaseq\_2016/main/qualimap</i>
<details>
<summary>:key: Click to see how to copy over the results, sample by sample</summary>
{% highlight bash %}
cp -r /sw/courses/ngsintro/rnaseq/main/qualimap/SRR3222410 $resDir/transcriptome/qualimap/
{% endhighlight %}
</details>
<details>
<summary>:key: Click to see how to copy over the results, using bash loop</summary>
{% highlight bash %}
for i in SRR3222409 SRR3222410 SRR3222411 SRR3222412 SRR3222413 SRR3222414
  do cp -r "/sw/courses/ngsintro/rnaseq/main/qualimap"/$i $resDir/transcriptome/qualimap/
done
{% endhighlight %}
</details>
<br />

:open_mouth: **Check the QualiMap results**. What do you think? Are the samples of good quality? How can you tell?  
<br />
<br />
[Jump to the top](#begin)
<br />
<br />


### <a name="featurecounts"></a> featureCounts: counting reads
After ensuring mapping quality we can count the reads to obtain a raw count table. We could count the reads by hand, opening the BAM in the IGV along the genome annotation file, and counting the reads overlapping with the regions of interest. This of course would take forever for the entire genome but it is never a bad idea to see how the data look like for the selected few genes of interest. For get the counts for the entire genome one can use many of the already available tools doing just that. Here we will use featureCounts, an ultrafast and accurate read summarization program, that can count mapped reads for genomic features such as genes, exons, promoter, gene bodies, genomic bins and chromosomal locations.

:mag: **Read** [featureCounts](http://bioinf.wehi.edu.au/featureCounts/) documentation and see if you can figure it out how to run summarize paired-end reads and count fragments overlapping with exonic regions.  
<br />

:computer: **Create featurecounts** sub-folder in the <i>transcriptome</i> directory and **navigate** there.
<details>
<summary>:key: Click to see how...</summary>
{% highlight bash %}
mkdir $resDir/transcriptome/featurecounts
cd $resDir/transcriptome/featurecounts
{% endhighlight %}
</details>
<br />

:computer: **Load featureCounts** module. :bulb: featureCounts is available on Uppmax as part of the <i>subread</i> package
<details>
<summary>:key: Click to see how to load featureCounts</summary>
{% highlight bash %}
module spider subread
module load subread/x.x
{% endhighlight %}
</details>
<br />

:computer: **Sym-link** SAM files generated by Star and located in <i>$resDir/transcriptome/star</i>
<details>
<summary>:key: Click to see sym-link files, sample by sample</summary>
{% highlight bash %}
ln -s $resDir/transcriptome/star/SRR3222409/SRR3222409_Aligned.out.sam
{% endhighlight %}
</details>
<details>
<summary>:key: Click to see sym-link files, using bash loop</summary>
{% highlight bash %}
for i in $resDir/transcriptome/star/**/*.sam
do 
  ln -s $i
done
{% endhighlight %}
</details>
<br />

:computer: **Run featureCounts** on the SAM files, **counting** fragments overlapping exon regions and **saving** the count tables as <i>tableCounts</i>. :bulb: The libraries are un-stranded and you can proceed all the samples in one go.
<details>
<summary>:key: Click to see how to run featureCounts on all samples</summary>
{% highlight bash %}
featureCounts -p -a $resDir/transcriptome/reference/Mus_musculus.GRCm38.85.gtf -t gene -g gene_id -s 0 -o tableCounts *.sam
{% endhighlight %}
</details>
<br />

:white_check_mark: Check if featureCounts run. You should have two files now:
{% highlight bash %}
ll -h $resDir/transcriptome/featurecounts/table*

-rw-rw-r-- 1 olga olga 2.8M Sep 15 11:05 tableCounts
-rw-rw-r-- 1 olga olga  658 Sep 15 11:05 tableCounts.summary
{% endhighlight %}
<br/>

:open_mouth: **Have a look** at the _tableCounts_ and _tableCounts.summary_. Can you figure out what these files contain? Do you think that counting work? How can you tell?
<br />
<br />
[Jump to the top](#begin)
<br />
<br />


### MultiQC: combining QC measures across all the samples
:mag: **Read** more on [MultiQC](http://multiqc.info). Can you figure out why this tool has become very popular? Can you figure out how to combine FastQC, Star, QualiMap and featureCounts results for all the samples into interactive report?
<br />

:computer: **Navigate** to _transcriptome_ directory and *load module MultiQC/0.6*
<details>
<summary>:key: Click to see how </summary>
{% highlight bash %}
cd $resDir/transcriptome
module spider MultiQC
module load MultiQC/x.x
{% endhighlight %}
</details>
<br />

:computer: **Run** MultiQC
<details>
<summary>:key: Click to see how to run MultiQC</summary>
{% highlight bash %}
multiqc .
{% endhighlight %}
</details>
<br />

:open_mouth: **Transfer** the MultiQC report to your computer and have a look at it.What can you notice?
<br />
<br />
[Jump to the top](#begin)
<br />
<br />


### <a name="descript"></a> Differential expression
As mentioned during the lecture, the best way to perform differential expression is to use one of the statistical packages, within **R environment**, that were specifically designed for analyses of read counts arising from RNA-seq, SAGE and similar technologies. Here, we will one of such packages called [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html). Learning R is beyond the scope of this course so we prepared basic ready to run scripts from a command line scripts to find DE genes between Ko and Wt.

:computer: **Create** <i>DE</i> sub-folder in the <i>transcriptome</i> directory and **navigate** there
<details>
<summary>:key: Click to see how </summary>
{% highlight bash %}
mkdir $resDir/transcriptome/DE 
cd $resDir/transcriptome/DE
{% endhighlight %}
</details>
<br />

:computer: **Load R module and R packages**. Match the R and R_package versions
<details>
<summary>:key: Click to see how </summary>
{% highlight bash %}
module spider R
module load R/x.y.z
module load R_packages/x.y.z
{% endhighlight %}
</details>
<br />

:computer: **Sym-link** <i>tableCounts</i> as created by featureCounts and **sym-link** the prepared gene annotation file *tableCounts_annotation.txt* prepared by us before the class and located: <i>/sw/courses/ngsintro/rnaseq/main/DE</i> as well as **sym-link** R script located in the same directory
<details>
<summary>:key: Click to see how </summary>
{% highlight bash %}
ln -s $resDir/transcriptome/featurecounts/tableCounts
ln -s /sw/courses/ngsintro/rnaseq/main/DE/tableCounts_annotations.txt
ln -s /sw/courses/ngsintro/rnaseq/main/DE/diffExp.R
{% endhighlight %}
</details>
<br />


:open_mouth: Open the diffExp.R script and quickly consider what of the above you would have to modify for your own samples
<br />

:open_mouth: Why do you think this requires so much more special considerations?
<br />

:computer: **Run the script**
<details>
<summary>:key: Click to see how </summary>
{% highlight bash %}
R CMD BATCH diffExp.R
{% endhighlight %}
</details>
<br />


:white_check_mark: A file <i>results\_DE.txt</i> should be created in the <i>DE</i> sub-folder  

{% highlight bash %}
ll -h $resDir/transcriptome/DE
..
-rw-rw-r-- 1 olga olga 8.9M Nov 29 16:31 DEdata.RData
lrwxrwxrwx 1 olga olga   50 Nov 29 16:30 diffExp.R -> /sw/courses/ngsintro/rnaseq/main/DE/diffExp.R
-rw-rw-r-- 1 olga olga 3.4K Nov 29 16:31 diffExp.Rout
-rw-rw-r-- 1 olga olga 2.6M Nov 29 16:31 results_DE.txt
lrwxrwxrwx 1 olga olga   55 Nov 29 16:30 tableCounts -> /proj/g2016035/nobackup/<username>/transcriptome/featurecounts/tableCounts
lrwxrwxrwx 1 olga olga   68 Nov 29 16:30 tableCounts_annotations.txt -> /sw/courses/ngsintro/rnaseq/main/DE/tableCounts_annotations.txt
{% endhighlight %}
<br/>

:open_mouth: **Copy over** to your computer **open** the *results\_DE.txt*
 file in Excel or alike. Given FDR value of 0.05, how many DE genes are there? How many up and down-regulated? What are the top changes? How does it change when we only look at the DE that have minimum log-fold-change 1?
<br />
<br />
[Jump to the top](#begin)

Another possibility would be to use <b>DEseq2</b> to perform the differential analysis, see <a href="https://scilifelab.github.io/courses/rnaseq/labs/DEseq2">https://scilifelab.github.io/courses/rnaseq/labs/DEseq2</a>
<br />
<br />

# Closing remarks and where to go next
It is not possible to learn RNA-seq data processing and analysis in one day... The good news is that there are many available tools and well-written tutorial with examples to learn from. In this tutorial we have covered the most important data processing steps that may be enough when the libraries are good. If not, there is plenty of trouble-shooting that one can try before discarding the data. And once the count table are in place, the biostatistical and data mining begins. There are no well-defined solutions here, all depends on the experiment and questions to be asked, but we strongly advise learning R. Not only to use the specifically designed statistical packages to analyze NGS count data, but also to be able to handle the data and results as well as to generate high-quality plots. There is no better way of learning than to try...

For those interested in RNA-seq analysis Scilifelab offer a more advanced course in RNA-sequnence analysis each semester. If you also have in interest in learning R we do for the first time this year offer a one-week introduction course in R programming. For more information on both of of these courses see [Courses offered by Scilifelab](https://www.scilifelab.se/education/courses/).
<br />
<br />
[Jump to the top](#begin)

# <a name="read"></a> More reading
- Robinson, MD, and Oshlack, A (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology 11, R25.


# About authors
Thomas KÃ¤llman :neckbeard:, Agata Smialowska :smiling_imp:, Olga Dethlefsen :angel: @ NBIS, National Bioinformatics Infrastructure Sweden
<br />
<br />
[Jump to the top](#begin)